packages/hydrogen/src/utils/request.ts

```diff
@@ -0,0 +1,46 @@
+import {generateUUID} from './uuid';
+
+export type CrossRuntimeRequest = {
+  url?: string;
+  method?: string;
+  socket?: {remoteAddress?: string}; // For Node request
+  headers: {
+    get?: (key: string) => string | null | undefined;
+    [key: string]: any;
+  };
+};
+
+export function getClientIp(request: CrossRuntimeRequest) {
+  return (
+    request.headers?.get?.('oxygen-buyer-ip') ??
+    request.headers?.get?.('cf-connecting-ip') ??
+    request.headers?.['x-forwarded-for']?.split(',')[0] ??
+    request.socket?.remoteAddress ??
+    null
+  );
+}
+
+export function getRequestId(request: CrossRuntimeRequest) {
+  let requestId = getHeader(request, 'request-id');
+
+  if (!requestId) {
+    // Store it in the headers object for later access
+    request.headers['request-id'] = requestId = generateUUID();
+  }
```

Comment:
This piece here is going to generate an ID and store it in the request headers object so that it can be used in the next call (e.g., when calling `createWithCache` and later `createStorefrontClient`, they should access the same request-id if the environment doesn't provide it).